### DISCUSSION TITLE
System won't start after upgrade if local plugin is enabled

### DISCUSSION THREAD

**Discussion by gunstr** [inventree/InvenTree#5066] [Q&A] [answered]
I have had the [IPN validation plugin](https://github.com/gunstr/inventree_ipn_validator) installed locally and running for a few month now.

The version information was:
InvenTree-Version: 0.11.0 dev
Django Version: 3.2.18
Commit Hash: 86f4a8294
Commit Date: 2023-03-21

Today I upgraded to the latest development version but then the system will not start if the plug in is enabled. The Supervisor status is OK:

![image](https://github.com/inventree/InvenTree/assets/41468206/ec1b4e2d-0c0a-4f4a-b18d-8be132002269)

But the system is not responding and if I look in the cluster.err.log the logging stops after `Loading plugin ValidateUniqueIPN` :

![image](https://github.com/inventree/InvenTree/assets/41468206/20bfcaf4-1355-4dac-9e99-3239e636b643)

If I disable the plugin before the upgrade the system starts as it should and the log also continues to be populated as below

![image](https://github.com/inventree/InvenTree/assets/41468206/4997036b-0702-4e7a-a143-be07e1768b06)

I know there has been recent changes related to plugins but what I can find nothing that should prevent local installation in the `src/InvenTree/plugins/` folder. I have spent the full day now to try to sort this out, but I cannot figure out what I have missed. Any suggestions?

**Comment by matmair** [comment]
It seems to run on master for me. Can you try a higher log level?

**Reply by matmair** [reply]
# Version Information:
InvenTree-Version: 0.12.0 dev
Django Version: 3.2.19
Commit Hash: 2e8fb2a
Commit Date: 2023-06-18
Commit Branch: master
Database: sqlite3
Debug-Mode: True
Deployed using Docker: False
Platform: macOS-13.4-arm64-arm-64bit
Installer: GIT

Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'ValidateUniqueIPN', 'slug': 'ipn_validator', 'version': '0.1'}, {'name': 'ShopifyPlugin', 'slug': 'shopify', 'version': '0.0.6a0'}]

**Comment by gunstr** [comment]
The log level is already set to DEBUG:
![image](https://github.com/inventree/InvenTree/assets/41468206/9fbb449a-ea4c-49ae-9a32-3251b1477331)

But I noticed another strange thing, after the upgrade I cannot open the "About" dialog... I get this error:

![Screenshot 2023-06-19 093900](https://github.com/inventree/InvenTree/assets/41468206/2882d954-e687-4b8b-98f9-b6ca87a2829c)

 Can it be related?

As a test I restarted again with a copy of my working system and deleted the plugin before i did the upgrade and I still get the "About" error above.

Full version info för my working copy:

# Version Information:
InvenTree-Version: 0.11.0 dev
Django Version: 3.2.18
Commit Hash: 86f4a8294
Commit Date: 2023-03-21
Database: mysql
Debug-Mode: True
Deployed using Docker: False
Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'Brother Labels', 'slug': 'brother', 'version': '0.5.0'}]

Then I upgrade to the master commit https://github.com/inventree/InvenTree/commit/2e8fb2a14ab84c73052d8a67c6c4d48a1cfd7a18

If I click the "show error information" there is a lot of info, but this is a screenshot of the beginning:

![image](https://github.com/inventree/InvenTree/assets/41468206/a33f1f27-f068-4941-82d5-adce2d5eb48f)

I have not a clue if this is related to the plugin problem and how to approach this...

Another thing I have noticed is a warning when I run 'invoke update':

Applying part.0109_auto_20230517_1048...System check identified some issues:
WARNINGS:
company.Address: (models.W036) MySQL does not support unique constraints with conditions.
        HINT: A constraint won't be created. Silence this warning if you don't care about it.

If you have any suggestions it's very appreciated.

**Reply by SchrodingersGat** [reply]
@gunstr thanks for repoting the additional errors - I have raised an issue for each. But, most likely not the cause of what you are seeing here.

**Comment by SchrodingersGat** [comment]
- Ref: https://github.com/inventree/InvenTree/issues/5071
- Ref: https://github.com/inventree/InvenTree/issues/5070

**Reply by matmair** [reply]
@gunstr We require python 3.9 or higher; [see here](https://github.com/inventree/InvenTree/blob/master/CONTRIBUTING.md#target-version).
removeprefix is a std lib function of str starting in 3.9. Please update your version of python as you are using an unsupported version (3.8.10)

**Comment by gunstr** [comment]
I have narrowed down the problem a bit. The last commit I can apply is https://github.com/inventree/InvenTree/commit/2ed7eefa27db09d7eb09544e6efd7b47fbefa37b.
If I apply the next commit https://github.com/inventree/InvenTree/commit/58a33c2e670881abe0e20e372b1d362f4d3dc2fc (Switch from git to dulwich) the system will not respond at all as long as the plug in is enabled.

**Comment by gunstr** [comment]
Sorry, I have missed the python 3.9 requirment :-(
I will upgrade, might solve the plugin problem also...

**Comment by gunstr** [comment]
Now I have upgraded to Python3.9.17 but the problem with loading the plugin is still there. It works fine up to commit https://github.com/inventree/InvenTree/commit/2ed7eefa27db09d7eb09544e6efd7b47fbefa37b but when I upgrade to https://github.com/inventree/InvenTree/commit/58a33c2e670881abe0e20e372b1d362f4d3dc2fc it leaves the system in the same state as described above.

This is before the upgrade:

![Screenshot 2023-06-19 155520](https://github.com/inventree/InvenTree/assets/41468206/350197a5-121b-48d8-8b61-9efd0c57074f)

Version Information:
InvenTree-Version: 0.12.0 dev
Django Version: 3.2.19
Commit Hash: 2ed7eefa2
Commit Date: 2023-06-05
Database: mysql
Debug-Mode: True
Deployed using Docker: False
Active plugins: [{'name': 'InvenTreeBarcode', 'slug': 'inventreebarcode', 'version': '2.0.0'}, {'name': 'InvenTreeCoreNotificationsPlugin', 'slug': 'inventreecorenotificationsplugin', 'version': '1.0.0'}, {'name': 'ValidateUniqueIPN', 'slug': 'ipn_validator', 'version': '0.1'}]


One thing I noted on the terminal log is that during 'invoke update' it stops for around half a minute each time it reaches the `Loading plugin ValidateUniqueIPN` and then it continues 'Loaded plugin ValidateUniqueIPN` - like a timeout. But when I boot the system it's unresponsive and the logging in the cluster.err.log stops completely (I have waited for at least 5min)  --> UPDATE - I just noticed that the logging in the cluster.err.log actually continues after some minutes, but the web page is still not responding after 20min now,

**Comment by gunstr** [comment]
@matmair When you got the plugin working with the latest master where did you install it? As I never managed to find out why it's not discovered when I point towards an "external" folder I have put it just under `src/InvenTree/plugins/`. Could that be a reason why I cannot get it to work and you can?

![image](https://github.com/inventree/InvenTree/assets/41468206/b4a63a48-6635-436b-b94e-aa3fe5e15351)

**Reply by matmair** [reply]
I have cloned it into plugins. You are probably missing an __init__.py at some level. You need one at every level till the file in which the plugin class is defined.

**Comment by gunstr** [comment]
I just cannot get my arms around what is going on here... I have before just copied the 'validate_ipn_unique' folder to the 'src/InvenTree/plugins/' and as the problem starts when applying the commit where Dulwich is added I thougt it might be so that I had to clone the complete git repo as you do. I tried but with the same result - as soon as I enable the plugin the InvenTree web pages do no longer load - just this little round circle. 

Do you know if there are any OS related prerequisites for Dulwich I might have missed? I'm running the system on a barebone Ubuntu 20.04 as a VM but it's all updated to latest status.

I found in the 'server.err.log' that it's actually never starting if the plugin is enabled. Below is a section of the log when booting with the plugin enabled. So there is obiously a timeout that happens. And if you see in the attached complete log it's just restarting over and over again - forever...

```
Loaded plugin `InvenTreeCoreNotificationsPlugin`
Loaded plugin `InvenTreeCoreNotificationsPlugin`
Loading plugin `ValidateUniqueIPN`
Loading plugin `ValidateUniqueIPN`
Loading plugin `ValidateUniqueIPN`
[2023-06-20 10:49:09 +0000] [918] [CRITICAL] WORKER TIMEOUT (pid:1049)
[2023-06-20 10:49:09 +0000] [918] [CRITICAL] WORKER TIMEOUT (pid:1057)
[2023-06-20 10:49:09 +0000] [918] [CRITICAL] WORKER TIMEOUT (pid:1058)
[2023-06-20 12:49:09 +0200] [1049] [INFO] Worker exiting (pid: 1049)
[2023-06-20 12:49:09 +0200] [1057] [INFO] Worker exiting (pid: 1057)
[2023-06-20 12:49:09 +0200] [1058] [INFO] Worker exiting (pid: 1058)
[2023-06-20 10:49:10 +0000] [918] [WARNING] Worker with pid 1049 was terminated due to signal 9
[2023-06-20 10:49:10 +0000] [1461] [INFO] Booting worker with pid: 1461
[2023-06-20 10:49:10 +0000] [918] [WARNING] Worker with pid 1058 was terminated due to signal 9
[2023-06-20 10:49:10 +0000] [918] [WARNING] Worker with pid 1057 was terminated due to signal 9
[2023-06-20 10:49:10 +0000] [1462] [INFO] Booting worker with pid: 1462
[2023-06-20 10:49:11 +0000] [1463] [INFO] Booting worker with pid: 1463
2023-06-20 10:49:11,395 INFO Loading SECRET_KEY from '/home/inventree/src/InvenTree/secret_key.txt'
2023-06-20 10:49:11,403 DEBUG Configuring database backend:
2023-06-20 10:49:11,403 INFO DB_ENGINE: django.db.backends.mysql
2023-06-20 10:49:11,403 INFO DB_NAME: inventree
```

[server.err.log - plugin enabled.txt](https://github.com/inventree/InvenTree/files/11799211/server.err.log.-.plugin.enabled.txt)
[server.err.log - plugin not enabled.txt](https://github.com/inventree/InvenTree/files/11799231/server.err.log.-.plugin.not.enabled.txt)

**Reply by SchrodingersGat** [reply]
From those logs, it looks like the server is resource constrained, and the workers are being killed off. Might be getting stuck in a loop?

@matmair any comments about the link to dulwich?

**Reply by matmair** [reply]
In my tests, it was actually much lighter than the previous function.

**Comment by gunstr** [comment]
@SchrodingersGat Seems correct, question is why...

The VM was configured with 1CPU and 4GB ram, this is the load with the plugin not enabled:
![HTOP plugin not enabled](https://github.com/inventree/InvenTree/assets/41468206/c19d5fd2-21df-49e0-aa64-40209cf713a0)

The same config with the plugin enabled:
![HTOP plugin enabled](https://github.com/inventree/InvenTree/assets/41468206/44357ef1-e7d7-4b0a-87f3-7b93eda8d09e)

Reconfigured with 4CPU and 8GB ram
![HTOP plugin enabled 4CPU and 8GB](https://github.com/inventree/InvenTree/assets/41468206/f8a5364b-dfc4-4416-8935-a9edb31a1e4b)

So although the VM was a bit under dimentioned it's not the root cause. As you suggest it's rather that it's gettgin stuck in a loop. Any suggestion how to identify where?

**Reply by wolflu05** [reply]
Have you already tested other plugins so we can sort out if it's specific to this plugin.

**Comment by matmair** [comment]
@gunstr can you try installing `python-dulwich` and report back?

**Comment by gunstr** [comment]
Tests I have done:

1. `sudo apt install python3-dulwich` -  still ends up in the same loop with the plugin enabled
2. Disabled the `validate_ipn_unique` installed from `src/InvenTree/plugins/ ` folder and enabled the `CustomValidator` in the samples folder. System works as it should - no loop.
3. Moved the `CustomValidator` from the samples folder to the `src/InvenTree/plugins/ `.  When the plugin is enabled the system goes into the same loop with WORKER TIMEOUTs during boot. If I revert back to commit https://github.com/inventree/InvenTree/commit/2ed7eefa27db09d7eb09544e6efd7b47fbefa37b it again works with no issues.

Conclutions:
1. The problem only happens when any of the two plugins are loaded from the `src/InvenTree/plugins/ ` folder --> likely not a problem with the plugin as such
2. @matmair has tested the same with no issues --> likely the root cause is in my installation or system configuration

**Comment by gunstr** [comment]
Continue testing....

1. Applied the 'dulwich' commit https://github.com/inventree/InvenTree/commit/58a33c2e670881abe0e20e372b1d362f4d3dc2fc again
2. None of the two plugins currently in 'src/InvenTree/plugins/' starts correctly
3. Moved back the `CustomValidator ` to the samples folder - can again be enabled with no issues
4. Copied `validate_ipn_unique.py` to the same samples forlder - does not work. Deleted the file again.
5. Commented out all code for the `CustomValidator ` in `validation_sample.py` and instead added the code from `validate_ipn_unique.py` - and now it works!!!

What is really the root cause I do not know... Both files have the same permissions, are coded with UTF-8 and Unix LF but some kind of difference between those two files are bringing the system into the endless loop if I do not just reuse the existing file. Could it be something low level git related that is not visible just looking at the actual files?

And when the same files are located in 'src/InvenTree/plugins/' it does not work at all. One difference between the two locations is obviously that the 'src/InvenTree/plugins/' is untracked by git - could be  an idication for something related to git also here?

**Reply by SchrodingersGat** [reply]
I think I can replicate this.

1. Take file `validation_sample.py` from the plugin sample directory
2. Drop it into `./InvenTree/plugins/`
3. Change the "slug" (so it does not conflict with the existing sample)
4. Reload server and attempt to activate new plugin
5. The server takes a *long* time to get back up (multiple minutes). Could easily be causing the worker to time out.

**Reply by SchrodingersGat** [reply]
Putting the plugin into the "external" plugins directory also has this issue. It seems to happen for any plugin, not just the validation sample. Some simple profiling shows that it is taking ~75 seconds to load a plugin from either the internal or external "plugins" directory.

**Reply by SchrodingersGat** [reply, marked as answer]
Issue opened here - https://github.com/inventree/InvenTree/issues/5081

**Reply by matmair** [reply]
I will look into this
